Feature/implement target all map reducers

MrBr opened the pull request.

MrBr committed:
Introduce context

MrBr committed:
Change mapReducer to handle targetAll action

MrBr committed:
Add tests

MrBr committed:
Extend context

MrBr committed:
Add context tests

vedrani commented on src/context.js:
```diff
@@ -0,0 +1,53 @@
+
+export function extendActionToTargetAllMapReducers(action) {
+  return setContextProp(action, [TARGET_ALL], true);
+}
```
Sorry, but I think we should remove it from redux-composers. It adds functionality in redux-composers that don't belong here. Idea about context should be used in Rio v1/v2 maybe, and should be extended to enable serialization of such objects.

MrBr replied:
All right, I'll remove it, I agree it doesn't belong here.

vedrani commented on src/mapReducers.js:
```diff
@@ -1,4 +1,5 @@
 import _ from 'lodash';
+import { isTargetAllAction } from './context';
```
I would expect to define exported const for key that is by convention defined as "apply action to all keys in state". 
Something like `all = "@@map-reducers/all"` and if keySelector returns such key, then mapReducers reducers must apply action to each key in state.

MrBr replied:
We must have reserved property for such action because we want to target all reducers with single action, and it is not possible to set keySelector values to proper value for all reducers with single action (we do not know what keySelectors exists).
There is also a case when you want to target a map reducers group, then value of keySelector could be checked to target all mapped reducers, but it is different then what am I solving with this PR.

vedrani replied:
It is responsibility of keySelector function to return ALL based on internal logic of action creator or key selector function. I agree about groups, maybe we could allow key selector to return arrays of keys also.

MrBr replied:
Do we consider that action as "natural" for map reducers? Action to target all mapped reducers.

I think yes, and for a natural behavior nothing should be implemented additionally.

It makes more sense to add option to disable `AllAction` then to enable it every time. It is a lot easier to know that you do not want your reducer to get affected by `AllAction` then to know if it should handle `AllAction`.

You can not predict who will trigger `AllAction` and with what parameters.

vedrani replied:
This discussion takes too long, we can comment live. I will just answer your last comment. 
Your reducer doesn't know about ALL, only mapReducer should know about it. Idea of mapReducer is to apply action with reducer only to substate that is under key returned by keySelector. That's it.

MrBr committed:
Rename context to actionOptions

MrBr committed:
Support key selector to target all reducers

- If key selector now returns TARGET_ALL_REDUCERS value all mapReducer
  or mapFactoryReducers in the given map will be triggered

MrBr committed:
Action options improvements

- Naming and comment

vedrani commented on src/actionOptions.js:
```diff
@@ -0,0 +1,61 @@
+import _ from 'lodash';
+
+/**
+ * Context is additional, reserved property for saving
```
Options

vedrani commented on src/actionOptions.js:
```diff
@@ -0,0 +1,61 @@
+  if (!_.isPlainObject(action)) {
+    throw Error(
+      `Invalid action, can not add action options to the ${typeof action}.` +
+     'Action must be a object!'
```
an

vedrani commented on src/actionOptions.js:
```diff
@@ -0,0 +1,61 @@
+  return _.get(actionOptions, TARGET_ALL_OPTION_KEY);
+}
+
+export function extendActionToTargetAllMapReducers(action) {
```
I don't know if we want breaking change, but this name should be shorter. `applyToAll`?

MrBr replied:
I will rename it and make it backward compatible.
Calling the old function will log deprecation warning.

vedrani commented on src/mapReducers.js:
```diff
@@ -33,16 +42,28 @@ export function mapReducerFactory(keySelector, reducerFactory) {
       [key]: reducers[key](state[key], action),
     };
   };
+  function getReducer(key) {
```
I would leave it as it was, without this function

MrBr committed:
Fix PR comments

zrumenjak commented:
This seems to invasive only to support one special case. I don't know what you discussed in person. But I would go with an approach where the key selector would be responsible for handling those situations.

MrBr commented:
What do you mean by the key selector? It is not possible in the current implementation of the key selector to handle all cases we have.

We need mechanism to bypass the key selector for **all** mapped reducers and "activate" them.

We do not know how developers will implement the key selector.

zrumenjak commented:
@MrBr the key selector should be responsible to return which reducers should be targeted. We can add a helper function to RIO that would return which actions are wildcard actions (i.e., they should target all reducers). That would allow us to maintain this list of actions in one place, but it would avoid adding all this complexity to each action...

MrBr commented:
The key selector should be and it is responsible for all but one case when we want to target all reducers. I think it is part of the `mapReducer` internal implementation to allow such behavior because it bypass the key selector.

We can not have some actions to be specifically related to target all because one action `type` can be used for different cases. Sometimes you want to target all sometimes you do not.

I mean I am not satisfied with the idea that we put something into `action.meta` but that is @vedrani request. He refers to the [Flux standard actions](https://github.com/acdlite/flux-standard-action) as obligatory in a way.

We could have specific `targetAll` action that would wrap actual action and `mapReducer` could extract it and pass it to the mapped reducers but then actual action is hidden for other not mapped reducers.

zrumenjak commented:
Can you first describe some real world use cases where you need this feature?

MrBr commented:
Cache invalidation.

When app is open from the background mode we want to check if all collections are still valid, `checkExpiration` action is called.

zrumenjak commented:
Ok, so you could handle this in your key selector very easily:
```
(action) => {
  if (action.type === CHECK_EXPIRATION) {
    return ALL_KEYS;
  }

  return action.meta.collection;
}
```
or, if we want to contain this logic in RIO:
```
rioKeySelector((action) => action.meta.collection);
```
Where `rioKeySelector` would first handle the case when an action should target all keys, or call the selector from the argument in other cases.

MrBr commented:
I am aware of that option, but as I mentioned above I think it is all right for the `mapReducers` to have special case where we bypass the key selector.

Look at it from a different angle for a bit, doesn't it have sense that `mapReducer` have this special case? If it has to be handled from the outside it is not the special case.

You are saying that we are too aggressive towards the action but it is not something we do randomly, developer explicitly extends the action with that option.

`applyToAll(someActionCreator(...params))`.

It is much simpler then to think about all possible cases upfront.

I don't think `mapReducer` is here to guard mapped reducers from all actions, it is here to create wanted subtree and this `guard` is just a by product.

zrumenjak commented:
This is handled from the outside either way. Your approach is to add this to the action, which is generic and can be handled by multiple reducers that may or may not be in a `mapReducers`.

On the other hand, this issue only happens when a reducer is mapped, and a key selector is a concept introduced by `mapReducers`. Because of all that, I would say that it is more reasonable to implement this in a key selector. This approach is also more flexible, because you can decide which actions will target all child reducers, and which will not on a per-reducer basis.

MrBr commented:
I understand the problem that we do not know what part of action will be used and how to create a new state. 

However if we depend exclusively on the key selector then there is no way to target all mapped reducers by random action introduced by third party. My only concern is that we want that.

We could have additional middleware which would dispatch two actions. One special to for mapped reducers and the original one. Middleware could be added to the chain only if support to target all mapped reducers is needed. Action wrapper would still be needed in that case.

vedrani commented:
@dodsky told me about a problem. I agree that we should apply action to all reducers if keySelector returns undefined or null.

zrumenjak commented:
@vedrani, we agreed on that in the end, because it would allow all existing actions to function as they normally would when there are no mapped reducers in the pipeline. Only actions with a property that the key selector targets will be redirected to a specific reducer, all other actions will be directed to all reducers.

I would only target all reducers when the selector returns `undefined`, `null` could then be used to suppress an action altogether.

Commit:
MapReducer now targets all keys if key selector returns undefined

Commit:
Version bump (invalid publish)

vedrani commented on src/mapReducers.js:
```diff
@@ -35,19 +43,19 @@ export function mapReducer(keySelector, reducer) {
       }, {});
     }
 
-    if (!key) {
```
I would still keep it as security measure

vedrani commented on src/mapReducers.js:
```diff
@@ -60,20 +68,20 @@ export function mapReducerFactory(keySelector, reducerFactory) {
       }, {});
     }
 
-    if (!key) {
```
Same here

vedrani reviewed (approved):
Add 'target all' description into README.md and mapReducer with option to pass reducer factory.

Commit:
Removed alpha, updated docs
Added key safety checks on map reducer
